### PR TITLE
Adds support for new downsampling UX

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

This commit updates the UX for creating downsampled data for Prometheus metrics.

Now, downsampling can be done by
```
metrics:
  downsampling:
  - interval: 5m
    retention: 30d
  - interval: 1h
    retention: 365d
```

The above dataset-config setting will create 2 downsampling, ds_5m (interval 5m & retention 30d) and ds_1h (interval 1h & retention 365d).

If an entry is removed, like
```
metrics:
  downsampling:
  - interval: 1h
    retention: 365d
```

Then we disable the ds_5m downsampling from refreshing. Deletion can only happen via SQL, `CALL _prom_catalog.delete_downsampling(<schema_name>)`. This keeps the UX clean and simple.

If the removed downsampling is applied again later on, we simply enable the refreshing.

This also means that this commit adds ability to enable or disable downsampling as well.

## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [ ] CHANGELOG entry for user-facing changes
- [ ] Updated the relevant documentation

### CODE REVIEW

File: pkg/runner/client.go
Comment (Harkishen-Singh) [inline pkg/runner/client.go:182]: Please ignore this. This was the result of a bad conflict resolution. We do not plan to apply downsampling if dataset-config and downsampling is explicitly given, in order to be in the safe place.

File: pkg/downsample/downsample.go
Comment (niksajakovljevic) [inline pkg/downsample/downsample.go:51]: I feel we can simplify this logic by merging two sets in memory: one is array of new `Config`, another is array of existing `Config`. Upon merge we just UPSERT Configs (insert or update existing). Having a `merge` method defined will also allow for easier unit testing.

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [inline pkg/internal/day/duration.go:24]: Is internal `text` being used anywhere? I did a quick scan and didn't find any usage of it.
If not being used I'd recommend avoid creating new Duration type.

**Review by niksajakovljevic — COMMENTED** [review]
Left few comments after quick pass. I haven't digged deep into tests since there are a lot of linter complaints.

File: pkg/internal/day/duration.go
Comment (Harkishen-Singh) [reply pkg/internal/day/duration.go:24]: Yes, we use it in `Config.Name()`. It is important, otherwise the downsampling schemas will be created as `ds_24h` if user provided `1d` as input.

**Comment by Harkishen-Singh** [comment]
Note for reviewers: The lint tests fails at the moment. I haven't fixed them since I aim to first be sure that **we are fine with the new way `day.Duration` is implemented**. Once we are fine, I will fix the linting tests. Otherwise, if we plan to change `day.Duration` down the line, it will consume good time updating the tests.

**Comment by Harkishen-Singh** [comment]
Note: Commit [Adds support to query default downsampling views in absence of __column__](https://github.com/timescale/promscale/pull/1808/commits/7464174d04281b2e81652fa8c1c6c08b13378075) is not directly related to this PR, but it was more comfortable and faster for me to include in this PR, than wait for approval and merge and then open a following PR.

File: pkg/pgmodel/querier/query_builder_samples.go
Comment (Harkishen-Singh) [inline pkg/pgmodel/querier/query_builder_samples.go:216]: See code around https://github.com/timescale/promscale_extension/pull/596/files#r1060609175 for more info.

File: pkg/downsample/downsample.go
Comment (alejandrodnm) [inline pkg/downsample/downsample.go:50]: Are there constraints regarding multiple configs for the same interval? I'm thinking about what happens if someone does `24h` and `1d`. In this case they are not the same interval because they differ on the text representation which is what's being used as the key here.

File: pkg/downsample/downsample.go
Comment (alejandrodnm) [inline pkg/downsample/downsample.go:27]: I don't see where this `disabled` flag is being used 🙈

File: pkg/downsample/downsample.go
Comment (alejandrodnm) [inline pkg/downsample/downsample.go:46]: You could refactor the `Sync` method into multiple functions per each step. Something like:

```
func Sync(newCfgs) {
  existingCfgs, err := getExistingResolutions()
  err = updateExistingConfigs(existingCfgs, newCfgs)
  err = disableRemovedConfigs(existingCfgs, newCfgs)
  err = enableNewConfigs(existingCfgs, newCfgs)
}
```

The same goes if you go the UPSERT route as niksa recommended

**Review by alejandrodnm — CHANGES_REQUESTED** [review]

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [reply pkg/internal/day/duration.go:24]: Yeah I was thinking the same. It seems the only way out is to always use `minute` granularity?

Comment (niksajakovljevic) [reply pkg/internal/day/duration.go:24]: Yeah. It seems we are better with just using Duration String(). For example: `time.Second * 3600` and `time.Minute * 60` will both print `1h0m0s` which is what we want.

**Review by jgpruitt — DISMISSED** [review]

File: pkg/internal/day/duration.go
Comment (Harkishen-Singh) [reply pkg/internal/day/duration.go:24]: > I asked it [here](https://github.com/timescale/promscale/pull/1808/files#r1060615105) but is there a problem if a user creates 2 configs, one with 24h and another with 1d?

Yes, that is a valid point. I am not sure how to work around this. Doing `duration.String()` is not good, since then we will have schemas like `ds_1h0m0s` which is not tidy. We cannot choose to convert everything into hours or minutes, otherwise it hurts readability. So I left this as the user gave input. Any other suggestions please?

File: pkg/downsample/downsample.go
Comment (Harkishen-Singh) [reply pkg/downsample/downsample.go:50]: This is a good point. I missed it. Maybe I should error after finding a similar resolution in the database.

**Comment by Harkishen-Singh** [comment]
This ([`_prom_catalog.create_or_update_downsampling`](https://github.com/timescale/promscale_extension/blob/7ef9ea79390bcd94716e6fbf4653bdea770a5a6a/migration/idempotent/017-metric-downsampling.sql#L233)) might be helpful while reviewing.

**Review by alejandrodnm — APPROVED** [review]
lgtm

File: pkg/downsample/downsample.go
Comment (niksajakovljevic) [inline pkg/downsample/downsample.go:60]: I believe this lock is used for other purposes as well. So I'd recommend waiting/blocking until lock is taken an configuration is applied. We can obviously add some timeout and error if we were unable to take lock and apply downsampling configuration.

File: pkg/downsample/downsample.go
Comment (niksajakovljevic) [inline pkg/downsample/downsample.go:20]: Frankly having three different SQL functions to update downsampling seems confusing

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [inline pkg/internal/day/duration.go:27]: This field name should be more descriptive. eg. `durationStr`

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [inline pkg/internal/day/duration.go:28]: This as well should be more descriptive, eg `duration`

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [inline pkg/internal/day/duration.go:83]: If fields are public we don't need function to access them

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [inline pkg/internal/day/duration.go:93]: Since this field is public we don't need a setter func

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [reply pkg/internal/day/duration.go:24]: >Doing duration.String() is not good, since then we will have schemas like ds_1h0m0s which is not tidy

Is that problematic in any way? Mostly asking b/c I want us to avoid Duration wrapper

File: pkg/internal/day/duration.go
Comment (niksajakovljevic) [inline pkg/internal/day/duration.go:93]: Also it feels that `Duration` should be immutable because it's derived from text input.

File: pkg/downsample/downsample.go
Comment (niksajakovljevic) [reply pkg/downsample/downsample.go:20]: I can hardly tell what is the difference between: `update_downsampling_state` and `set_downsampling_state`. Also there is another function named `create_or_update_downsampling`.

**Comment by niksajakovljevic** [comment]
>Note: Commit [Adds support to query default downsampling views in absence of __column__](https://github.com/timescale/promscale/pull/1808/commits/7464174d04281b2e81652fa8c1c6c08b13378075) is not directly related to this PR, but it was more comfortable and faster for me to include in this PR, than wait for approval and merge and then open a following PR.

Btw it does seem that commit is gone so can't really follow through that part

File: pkg/downsample/downsample.go
Comment (jgpruitt) [reply pkg/downsample/downsample.go:50]: is this correct?

1. if existing config is missing in new set, disable it
2. if new config is not existing, insert
3. if new config matches existing config on schema name, update it UNLESS the interval is different, in which case, throw an error
4. no two configs can have the same interval
5. no two configs can have the same name

File: pkg/internal/day/duration.go
Comment (jgpruitt) [reply pkg/internal/day/duration.go:24]: can the user do `1.5d` and `36h`?

Comment (jgpruitt) [reply pkg/internal/day/duration.go:24]: I believe this will produce a desired standardized output regardless of how the user specifies the input.

here is the output for the samples given:

```
1d
1d12h
12h
12h15m
15m30s
1d12h15m30s
```

```go
package main

import (
	"fmt"
	"time"
)

func main() {
        // sample tests
	fmt.Println(String(time.Hour * 24))
	fmt.Println(String(time.Hour * 36))
	fmt.Println(String(time.Hour * 12))
	fmt.Println(String(time.Hour*12 + time.Minute*15))
	fmt.Println(String(time.Minute*15 + time.Second*30))
	fmt.Println(String(time.Hour*36 + time.Minute*15 + time.Second*30))
}

func String(d time.Duration) string {
	const day = int64(time.Hour * 24)

	remainder := int64(d)
	days := remainder / day
	remainder = remainder % day
	hours := remainder / int64(time.Hour)
	remainder = remainder % int64(time.Hour)
	mins := remainder / int64(time.Minute)
	remainder = remainder % int64(time.Minute)
	secs := remainder / int64(time.Second)

	display := ""
	if days != 0 {
		display = fmt.Sprintf("%dd", days)
	}
	if hours != 0 {
		display = fmt.Sprintf("%s%dh", display, hours)
	}
	if mins != 0 {
		display = fmt.Sprintf("%s%dm", display, mins)
	}
	if secs != 0 {
		display = fmt.Sprintf("%s%ds", display, secs)
	}
	return display
}
```

Comment (jgpruitt) [reply pkg/internal/day/duration.go:24]: We need to have a deterministic algorithm for how we name schemas. There may be situations (hopefully not) where we have to reverse engineer schema names to figure out what they were.

Comment (jgpruitt) [reply pkg/internal/day/duration.go:24]: So we could use the standard library's `time.Duration` for parsing, and use the function above (or similar) for generating the schema name.

Comment (jgpruitt) [reply pkg/internal/day/duration.go:24]: my vote: if we want to support `d` units in input, use `duration.Day` BUT use a standardize output format like the above instead of the raw user input.

Comment (jgpruitt) [reply pkg/internal/day/duration.go:24]: If we have a deterministic algorithm for the schema names, then we don't have to worry about detecting the case where two different schema names have the same interval. That's a big win IMO

File: pkg/downsample/downsample.go
Comment (jgpruitt) [reply pkg/downsample/downsample.go:50]: honestly, I'd write a single database function to handle this. It's much fewer lines of code and one database round-trip.

something like this:
```sql
create or replace function apply_downsample_config(config jsonb) returns void
as $proc$
declare
    _input _prom_catalog.downsample[];
    _schema_name text;
begin
    select array_agg(x) into strict _input
    from jsonb_populate_recordset(null::_prom_catalog.downsample, config) x;

    for _schema_name in
    (
        select x.schema_name
        from unnest(_input) x
        where not exists
        (
            select 1
            from _prom_catalog.downsample d
            where d.schema_name = x.schema_name
        )
    )
    loop
        execute format('create schema %I', _schema_name);
    end loop;

    -- insert or update
    insert into _prom_catalog.downsample (schema_name, resolution, retention, should_refresh)
    select x.schema_name, x.resolution, x.retention, true
    from unnest(_input) x
    on conflict (schema_name) do
    update set
        retention = excluded.retention,
        should_refresh = true
    ;

    -- disable if missing
    update _prom_catalog.downsample set should_refresh = false
    where schema_name not in 
    (
        select schema_name
        from unnest(_input)
    );
end;
$proc$ language plpgsql;
```

then you can just push the new config to the database as json. almost no logic required on the go side

```sql
select apply_downsample_config(
$$
[
    {
        "schema_name": "d_1d",
        "resolution": "1d",
        "retention": "30d"
    },
    {
        "schema_name": "d_12h",
        "resolution": "12h",
        "retention": "18d"
    },
    {
        "schema_name": "d_1h",
        "resolution": "1h",
        "retention": "15d"
    }
]
$$::jsonb
);
```

**Review by jgpruitt — CHANGES_REQUESTED** [review]
I strongly advocate for deterministic schema names.
I'd also like to see the simplified logic for updating the downsample config table per my comments

File: pkg/downsample/downsample.go
Comment (jgpruitt) [reply pkg/downsample/downsample.go:20]: If we named the setting `enable_downsampling` instead of `downsample` and then named the functions `set_enable_downsampling` and `get_enable_downsampling` that would clear things up a little IMO

File: pkg/downsample/downsample.go
Comment (Harkishen-Singh) [reply pkg/downsample/downsample.go:50]: I am not confident for pushing down the entire config for 3 reasons:
1. The point of connector applying the configuration is that in future, if the `downsample.Config` needs more fields, we can easily change there and connector handles it. With `apply_downsample_config`, I don't see how we can do unless we make changes in the extension code.
2. This can be specific to me. I wrote SQL code so it's easy to read and understand (I am not an expert in SQL). In the current SQL code, I tried to keep everything very straightforward since for me, I find following SQL not as easy as Golang. I am thankful for the new suggestion, but I think `apply_downsample_config` is not a straightforward function. Plus, if there happens to be any changes in `downsample.Config` in future, we cannot simply push the jsonb until we filter some fields, since the `_prom_catalog.downsample` will not have the additional fields in future. I feel this is creating a strong coupling.
3. With `apply_downsample_config()`, the SQL API is no longer simple. It does more than 1 task and I am not sure if its fine.

These are my thoughts. I do not want to oppose it, as I feel I am not the right person in SQL to suggest. So if everyone is happy with `apply_downsample_config` then I can write this function.

Comment (alejandrodnm) [reply pkg/downsample/downsample.go:50]: My 2 cents. I like the approach of the DB function:

1- If we add more fields to the config we would still need to store them in the DB, right?. This is just to handle the logic of applying the whole config. I find this simpler than doing the whole code in the connector and managing the logic there. For me, it feels less error prone and doesn't seem like an out of place extension to the downsample API.

2- 
> I am not an expert in SQL

This is precisely why you should try to write it. You'll have the guidance of the awesome data-platform team. They'll review and help you, and get you a little bit closer to being a SQL expert.

If you don't feel confortable working with the JSONB because of the loose schema, maybe there are alternatives. Could we have the config be a custom type and just pass an array of those, or something, just an idea I'm not a PostgreSQL expert either 😅 

3- You could say that this function does 1 thing, it replaces the current config with the given config.

Comment (Harkishen-Singh) [reply pkg/downsample/downsample.go:50]: Sorry, my point was more towards keeping things simple and straightforward. That will make life easier for all of us. Plus, keeping things in Golang allows us to test easily.

Comment (Harkishen-Singh) [reply pkg/downsample/downsample.go:50]: There are more points than just these. If we make create/update/disable in extension, then I will have to remove `TestMetricDownsampleSync` completely (and write in SQL). Otherwise, it will be creation/updation in extension, and its tests in Go. Then, I will have to write tests in SQL and it will be a very big PR to review. Given we are not sure of this feature's impact ATM, I don't feel confident investing more time and designing more. We should probably go now with speed than accuracy.

File: pkg/downsample/downsample.go
Comment (niksajakovljevic) [inline pkg/downsample/downsample.go:83]: Do we need both `Close` and `Unlock`? I'd assume that `Close` will unlock

**Review by niksajakovljevic — APPROVED** [review]

**Review by jgpruitt — APPROVED** [review]

**Pull request merged**
